### ISSUE TITLE
Do we need the gtk+ backend?

### ISSUE THREAD

**Issue by bjourne** [factor/factor#1932]
On Linux,we have two UI backends for Factor; gtk and x11. The gtk backend is default, but you can switch to x11 by loading the `ui.backend.x11` vocab. They work the same but the x11 one is a bit faster than gtk. In the future we should switch to x11 and ditch the gtk backend. Then the ui wouldn't need to load the annoying gtk libraries like `libgtk-x11-2.0.so`.

**Comment by jonenst** [comment]
Hmmm the gtk backend is supposed to be more robust and recent. If we had to
ditch one, I would think it's the x11 one. For example the icon doesn't
work with X11.
When do you notice a speed difference?

Le 4 févr. 2018 22:48, "Björn Lindqvist" <notifications@github.com> a
écrit :

On Linux,we have two UI backends for Factor; gtk and x11. The gtk backend
is default, but you can switch to x11 by loading the ui.backend.x11 vocab.
They work the same but the x11 one is a bit faster than gtk. In the future
we should switch to x11 and ditch the gtk backend. Then the ui wouldn't
need to load the annoying gtk libraries like libgtk-x11-2.0.so.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/factor/factor/issues/1932>, or mute the thread

**Comment by catb0t** [comment]
I don't think that it's a good idea, long-term, to drop support for GTK+ (2 or the possibility of 3) and make the X binding really good -- X on modern Linux desktops will die at some point in the near future, to be replaced by Weston or another Wayland implementation. 

At that future point, the  X code might still work with XWayland, but Wayland (implementations) have direct and good support for GTK3 (mostly because Canonical decided to use GNOME on Wayland for Ubuntu 17.10+).

GTK2 will continue to exist in use and maintained, in my opinion probably longer than X will stay relevant. Even if it is only preserved by the MATE project (the maintained fork of GTK2), that's okay for Factor.

**Comment by jonenst** [comment]
On 18.04, Ubuntu uses X.org by default, but still the Wayland point is valid.

**Comment by catb0t** [comment]
https://developer.gnome.org/gtk3/stable/gtk-migrating-2-to-3.html

If you (or @erg)  are interested in really porting to GTK+3 and want to do it soon (it's not really urgent), here are some of the things that should be looked at

Preparation
- [x] Do not include individual headers (N/A)
- [ ] Do not use deprecated symbols
- [ ] Use accessor functions instead of direct access
- [ ] Replace GDK_<keyname> with GDK_KEY_<keyname>
- [x] Use GIO for launching applications (N/A)
- [x] Use cairo for drawing (Factor already does this)

Switching 
- [ ] Replace size_request by get_preferred_width/height
- [x] Replace GdkRegion by cairo_region_t
covered by already using Cairo 
- [x] Replace GdkPixmap by cairo surfaces
covered by already using Cairo 
- [ ] Replace GdkColormap by GdkVisual
- [ ] GdkDrawable is gone
- [ ] Event filtering
- [ ] Backend-specific code
"In GTK+ 3, GDK can be built with multiple backends, and currently used backend has to be determined at runtime"
Factor only uses GTK on Linux / non-mac Unix so this shouldn't be an issue
- [x] GtkPlug and GtkSocket
N/A: X11-specific UI features which are now X11+GTK-only
- [ ] The GtkWidget::draw signal
- [x] GtkProgressBar orientation
Factor doesn't implement progress bars (yet?)
- [ ] Check your expand and fill flags
These flags are now inherited from containers
- [ ] Scrolling changes
"default values for the “hscrollbar-policy” and “vscrollbar-policy” properties have been changed from 'never' to 'automatic'"
- [ ] GtkObject is gone
- [ ] GtkEntryCompletion signal parameters
- [ ] Resize grips
"The resize grip functionality has been moved from GtkStatusbar to GtkWindow"
- [ ] Prevent mixed linkage
Especially careful if we decide to keep GTK 2 and 3
- [ ] Install GTK+ modules in the right place

**Comment by bpollack** [comment]
Only about half that list applies to us (a lot of it assumes we're directly consuming the C headers, which we're not), but it's probably worth using. NB to anyone who follows, the above list is literally the top hit for "upgrade gtk2 to gtk3," so it might be worth tracking that just in case the recommended advice changes.

**Comment by xerz-one** [comment]
Hi, I wonder, any progress on this? GTK4 is about to come out (3.99 is already available) and I would very much appreciate HiDPI support for my Surface. Thanks in advance!

**Comment by mrjbq7** [comment]
Ported to GTK3.

Working towards GTK4, probably get to it sometime in January.

**Comment by theCapypara** [comment]
Would it be possible to get a release with the GTK4 port soon-ish?

factor currently relies on GTK2 and gtkglext (abandoned since 2014), which are both pending removal from the NixOS distribution.

Edit: Not asking for this to be pushed, just wanting to make aware of that, we'll have to drop Factor potentially if it isn't updated.

**Comment by mrjbq7** [comment]
We now use GTK3

https://re.factorcode.org/2025/12/migrating-to-gtk3.html

It is available in the development branch although the nightly build currently is a little stale. 

We hope to release 0.102 in the next month with it. 

Is GTK3 a problem also? We wanted to do that and then do GTK4 after.

**Comment by theCapypara** [comment]
Oh sorry, no GTK3 is fine too, I somehow thought you ported straight to GTK4. 

Thanks!
